src/metabase/channel/render/body.clj

```diff
@@ -388,11 +389,138 @@
       [:img {:style (style/style {:display :block :width :100%})
              :src   (:image-src image-bundle)}]]}))
 
+(defn- get-col-by-name
+  [cols col-name]
+  (->> (map-indexed (fn [idx m] [idx m]) cols)
+       (some (fn [[idx col]]
+               (when (= col-name (:name col))
+                 [idx col])))))
+
+(defn- extract-value-sources
+  "Extracts column references from mappings that aren't strings"
+  [mappings]
+  (->> mappings
+       vals
+       (apply concat)
+       (filter (complement string?))))
+
+(defn- source-is-name-ref?
+  "Checks if value is a data source name reference"
+  [value]
+  (and (string? value)
+       (str/starts-with? value "$_")
+       (str/ends-with? value "_name")))
+
+(defn- value-source->card-id
+  "Parses the card id as an int out of a visualizer value source map
+   e.g. {:sourceId 'card:191', ...} -> 191"
+  [{:keys [sourceId]}]
+  (Integer/parseInt (second (str/split sourceId #":"))))
+
+(defn- name-source->card-id
+  "Extracts card id from name source string"
+  [value]
+  (try
+    (when-let [matches (re-find #":(\d+)_" value)]
+      (Integer/parseInt (second matches)))
+    (catch Exception _
+      nil)))
```

Comment:
What exception is possible here? `(second (re-find ...))` should return `nil` anyway if there's no match